2017/proposals/bill-engels-pymc3-proposal.md

```diff
@@ -0,0 +1,292 @@
+
+### Motivation
+
+PyMC3 relies on Theano to execute the computation graph and its gradient.
+However the vast majority of PyMC3's codebase is devoted to facilitating
+probabilistic modeling.  It includes libraries of inference methods, random
+variable types, methods to handle and analyze traces, and plotting
+functionality --- routines that are weakly bound at most to the computational
+backend.  Implementing an abstraction layer between PyMC3 core code and the
+computational backend will make the codebase more maintainable.  It would be
+easier to categorize issues as being related to either Theano or PyMC3 core
+code.  PyMC3 already has `theanof.py` and `math.py`.  Instead of making calls
+throughout the codebase directly to Theano, PyMC3 would interact with Theano
+through calls to the interface.
+
+A different backend may increase PyMC3's computational efficiency.  Results from
+benchmarking different frameworks must be taken with a big grain of salt in the
+context of PyMC3, since they naturally use deep neural networks and not
+probabilistic models as test cases.  The only way to find out would be to test!
+
+### Computational Backends
+
+The two most relevant deep learning frameworks to PyMC3 are PyTorch and
+TensorFlow (not counting Theano), based on their more extensive libraries of
+mathematical and linear algebra operations.
+
+The operations currently available are:
+
+- [Theano][theano-ops]
+- [TensorFlow][tf-ops]
+- [PyTorch][torch-ops]
+
+1. [Theano][theano-technote] represents a series of computations expressed by a
+user as a directed acyclic graph (DAG).  The DAG representation allows for
+speed increases via symbolic simplifications (such as $\frac{cx}{x} \rightarrow
+c$) and dynamic C code generation.  Some operations have methods defined on
+both the CPU and GPU.
+
+2. [TensorFlow][tensorflow-technote] generates a DAG in a similar fashion as
+Theano.  TensorFlow is a much broader framework than Theano.  It includes
+minimization routines, facilities for reinforcement learning, and it is
+designed to allow computation to scale on distributed systems, multiple GPUs,
+and specialized devices like Google's new TPUs.  The recent release of
+[XLA][xla] aims to optimize lower-level mathematical and linear algebra
+operations to improve execution speed.  TensorFlow appears to currently have
+the largest userbase of the three, based on a superficial look at GitHub
+activity.
+
+3. [PyTorch][pytorch-technote] is the new Python API to Torch.  Unlike Theano and
+TensorFlow, the PyTorch execution graph is dynamically defined.  Reverse-mode
+auto-differentiation is used to compute gradients on the fly.  Without the
+definition of a static computation graph before execution, its not possible for
+PyTorch to do symbolic optimizations like Theano and TensorFlow.  However, its
+dynamic computation graphs allows variable-length inputs and outputs to be
+processed.  This may make the implementation of prediction methods in PyMC3
```

Comment:
It might also allow models to have dynamic dimensionality during sampling which could be of interest for e.g. non-parametrics. I'm not sure if it's true, but if the execution graph is dynamic, does PyTorch handle if-else better? If so, the samplers could also be sped up quite a bit with this.